### DISCUSSION TITLE
Some questions about the architecture of pnpm?

### DISCUSSION THREAD

**Discussion by lancejpollard** [pnpm/pnpm#6654] [Q&A]
Hi there, loving pnpm as an alternative to npm/yarn which I've used in the past. I'm curious to learn more about how it is architected. I have read these so far:

- https://pnpm.io/motivation
- https://pnpm.io/symlinked-node-modules-structure
- https://pnpm.io/blog/2020/05/27/flat-node-modules-is-not-the-only-way
- https://pnpm.io/how-peers-are-resolved

I don't quite understand the symlink article yet, and don't know how the content addressable store is organized yet (need to find that out), but I have a few questions to start.

> If you depend on different versions of the dependency, only the files that differ are added to the store. For instance, if it has 100 files, and a new version has a change in only one of those files, pnpm update will only add 1 new file to the store, instead of cloning the entire dependency just for the singular change.

First, how does it do that? It's not using git I would assume, so how does it know only 1 file changed out of 100, it would have ot compare each file to every other file it seems, which seems like a lot of processing.

> All the files are saved in a single place on the disk.

Where are they saved? I see this in my project, but it looks like there is a local hidden directory with a copy of every module, so I don't see how you can say if 100 projects use the same modules, there is only 1 copy, what am I missing? I would assume there is some global folder somewhere like `~/.pnpm` where all the packages are symlinked from, but it does not seem to be the case.

<img width="1249" alt="Screenshot 2023-06-08 at 1 36 45 AM" src="https://github.com/pnpm/pnpm/assets/28718/b954492a-b036-48c7-a84b-16e6474085e7">

Lastly for now, why does it go `.pnpm/@babel+helper-string-parser@7.21.5/node_modules/@babel/helper-string-parser/<code>`? Why is there that `node_modules` folder below the versioned thing, and each symlink from the project node_modules folder links to like `.pnpm/@babel+helper-string-parser@7.21.5/node_modules/@babel/helper-string-parser`? That seems strange.

Thank you for the help.

**Comment by lancejpollard** [comment]
I found the content addressable store I think, at `~/Library/pnpm` on a Mac.

<img width="1264" alt="Screenshot 2023-06-08 at 1 49 44 AM" src="https://github.com/pnpm/pnpm/assets/28718/6b858ec9-e03f-469e-83ff-4cc1d88c690d">

It looks like you hash the files some how, and maybe that is how you can check to only add the ones that change? How does that work exactly?

But where are the content addressable _packages_? I only see the files?

Also, why does it say `global/5`? Why 5?

**Comment by RexSkz** [comment]
The first question leads to the CAS (content-addressable store), you can treat it as an object storage (just like S3) but "key" is the hash result of "content". So if you'd like to save a file with the hash `xxx` but there's already a file named `xxx` in CAS, you immediately know the file doesn't change.

Files in CAS will be hard-linked to your project's `node_modules/.pnpm` folder. "Hard link" is a link that looks like a normal "file".

A file is just a link to a piece of data in your hard drive, which means the file `/Library/pnpm/store/v3/files/00/945...` and `my-project/node_modules/.pnpm/xxx/yyy/package.json` (or another possible file) point to the same position in your hard drive though they look like two files.

So there's no "package", only "hard link" (or "file").

**Reply by lancejpollard** [reply]
So my question related to this is, how does it know how to reconstruct `my-project/node_modules/.pnpm/xxx/yyy/package.json` from `/Library/pnpm/store/v3/files/00/945...`? I don't see anywhere in the file structures where it has the information to tell it to convert `/Library/pnpm/store/v3/files/00/945...` to the specific path to "recreate" the package layout. It seems like in CAS it would have to have a manifest.json or something, to say how the files are laid out in a package. I saw some `.json` files in like `/Library/pnpm/store/v3/files/00/945...json`, but even then, how would it find these json files to know which package it's for? Can you please elaborate on that part? Thank you for your help!

**Reply by RexSkz** [reply]
It takes several steps. I'll use one of my packages `json-diff-kit@1.0.14` as an example:

First, get package info from https://registry.npmjs.org/json-diff-kit/1.0.14, the `dist.integrity` is `sha512-K6exI1H8...`, this is a base64 format, you can convert it to hex format `2ba7b123...` using https://base64.guru/converter/decode/hex.

Next, use this hex to find the index file (manifest): `[store]/v3/files/2b/a7b123...-index.json`, it shows all the files needed (just ignore the files which should be npm-ignored, I'll fix it later):

```json
{
  "name": "json-diff-kit",
  "version": "1.0.14",
  "files": {
    ".editorconfig": {
      "checkedAt": 1682391943134,
      "integrity": "sha512-3Araasa9Jnn6DwVZpe7nvi3IgopUNXzy5OyL3D2Dta1cbCnaC3WOCn7H/IZNpKPMus8kOSwC5msrqNBeE7+dSQ==",
      "mode": 420,
      "size": 165
    },
    "LICENSE": {
      "checkedAt": 1682391943133,
      "integrity": "sha512-wd09AhFHsJ3ZdkLH52q+EGmif1N/au6OMDIrxnlO4VGphQIUKwgzNNChcrHtw3Sflu3R5HfmdmglfhaDZUELHQ==",
      "mode": 420,
      "size": 1065
    },
    "coverage/lcov-report/base.css": {
      "checkedAt": 1682391943133,
      "integrity": "sha512-3cnYSyOC6wWEiz6+OA7lnjlNeb3lKGzfqNn7npS0Cx4bG5bjbpmXIoeNMxvGQe14fNKOcgDGiYQEnhlcALzYmw==",
      "mode": 420,
      "size": 5394
    },
    "coverage/lcov-report/prettify.css": {
      "checkedAt": 1682391943133,
      "integrity": "sha512-T6HWikKb8QrBXMTJu36ZjPe3Hrsv/Nk38WGTycimZ85LuVqQAn4djkuYVknhBZzLxdAML2neMUEXPvkJ9fzdww==",
      "mode": 420,
      "size": 676
    },
    "dist/viewer.css": {
      "checkedAt": 1682391943134,
      "integrity": "sha512-0bFxWz4jxMX3L26ebkGhoPLhYIJ7BkeDkiONhSYTt5dF9F5yrYDaThPNjCPzLSjDbDvj5Fyf073sGIzMTrk4YA==",
      "mode": 420,
      "size": 1889
    },
    ...
  }
}
```

Just look at the first file `.editorconfig`, it has the hash `sha512-3Araasa9...`, which is `dc0ada6a...` in hex, so let's cat the file `[store/v3/files/dc/0ada6a...]` (without `-index.json`):

```editorconfig
root = true

[*]
insert_final_newline = true
charset = utf-8
trim_trailing_whitespace = true
end_of_line = lf

[*.{ts,js,json}]
indent_style = space
indent_size = 2
```

**Reply by likepeach789** [reply]
Thank you so much for this detailed answer! : )  I follow the steps and did one test.
Regarding the saying`First, get package info from https://registry.npmjs.org/json-diff-kit/1.0.14, the dist.integrity is sha512-K6exI1H8..., this is a base64 format, you can convert it to hex format 2ba7b123...` 
Per My test, it is using `dist.shasum` rather than `dist.integrity`. btw, my pnpm version is V7.

**Comment by RexSkz** [comment]
The answer to the second question is that PNPM uses symlinks (soft links) to build the `node_modules` directory. It's a bit complicated to start over, let's just see the result:

- Node.js's module resolution algorithm will follow symlinks and return the real path. The real path will be at `my-project/node_modules/.pnpm/[package-name]@[version]/node_modules/[package-name]/[maybe index.js]` (remember it points to the same position as a file in CAS).
- If there's a `require('xxx')` in this file, Node.js will find `xxx` in these folders one by one:
  - `my-project/node_modules/.pnpm/[package-name]@[version]/node_modules/[package-name]/node_modules/xxx`
  - `my-project/node_modules/.pnpm/[package-name]@[version]/node_modules/node_modules/xxx`
  - `my-project/node_modules/.pnpm/[package-name]@[version]/node_modules/xxx` [situation 1]
  - `my-project/node_modules/.pnpm/node_modules/xxx`
  - `my-project/node_modules/node_modules/xxx`
  - `my-project/node_modules/xxx` [situation 2]
- Situation 1 is that `[package-name]` has a peer dependency `xxx` but there're multiple versions of `xxx`, PNPM uses an extra `node_modules` to bind the right version of `xxx` with `[package-name]`.
- Situation 2 is that `[package-name]` doesn't have a peer dependency `xxx`.
- Finally, the files in `xxx` are also symlinks to `node_modules/.pnpm`, which means Node.js has to follow symlinks again to find the real path.

**Reply by lancejpollard** [reply]
I need to better understand how the node resolution algorithm works, any tips on where a good description is of that? I am literally trying to figure out how to design a package manager for a custom language, and pnpm seems to have done it best! But I still have some gaps in my understanding.

**Reply by lancejpollard** [reply]
Specifically, does the node resolution algorithm first check for a package.json at each level? To then check the version of the package.json dep if present, which could be a range, and then given a range like `1.2.x`, how does it find the `1.2.3` version below, does it have to iterate through all folders in the node_modules directory to check? There's a lot of details in her I hope we can unpack.

**Reply by lancejpollard** [reply]
https://github.com/browserify/resolve/blob/main/lib/sync.js#L117

**Reply by RexSkz** [reply]
Here's the official document of Node.js's resolution algorithm: https://nodejs.org/api/modules.html#loading-from-node_modules-folders

> For example, if the file at '/home/ry/projects/foo.js' called require('bar.js'), then Node.js would look in the following locations, in this order:
>
> - /home/ry/projects/node_modules/bar.js
> - /home/ry/node_modules/bar.js
> - /home/node_modules/bar.js
> - /node_modules/bar.js

As for "which version it uses", it's determined by symlinks. The `my-project/node_modules/json-diff-kit` is a symlink to the folder in `my-project/node_modules/.pnpm`:

```
$ ls -l node_modules | grep json-diff-kit
lrwxr-xr-x  1 rexzeng  staff  53  6  9 16:16 json-diff-kit -> .pnpm/json-diff-kit@1.0.14/node_modules/json-diff-kit
```

So if there's a `import xxx from 'json-diff-kit'` in `my-project`, it imports the `1.0.14` version.

Actually, PNPM will link it to the version specified in `my-project/package.json`.
